nixos/modules/services/network-filesystems/ipfs-cluster.nix

```diff
@@ -0,0 +1,123 @@
+{ config, lib, pkgs, options, ... }:
+with lib;
+let
+  cfg = config.services.ipfs-cluster;
+  opt = options.services.ipfs-cluster;
+
+  # secret is by envvar, not flag
+  initFlags = toString [
+    (optionalString (cfg.initPeers != [ ]) "--peers")
+    (lib.strings.concatStringsSep "," cfg.initPeers)
+  ];
+in {
+
+  ###### interface
+
+  options = {
+
+    services.ipfs-cluster = {
+
+      enable = mkEnableOption
+        "Pinset orchestration for IPFS - requires ipfs daemon to be useful";
+
+      user = mkOption {
+        type = types.str;
+        default = "ipfs";
+        description = "User under which the ipfs-cluster daemon runs";
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "ipfs";
+        description = "Group under which the ipfs-cluster daemon runs";
+      };
+
+      consensus = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description = "Consensus protocol - 'raft' or 'crdt'";
+      };
+
+      dataDir = mkOption {
+        type = types.str;
+        default = "/var/lib/ipfs-cluster";
+        description = "The data dir for ipfs-cluster";
+      };
+
+      initPeers = mkOption {
+        type = types.listOf types.str;
+        default = [ ];
+        description = "Peer addresses to initialize with on first run";
+      };
+
+      secret = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description =
+          "Secret for an existing cluster; if null, a new secret is generated";
+      };
+
+      secretFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description =
+          "File containing the secret - 'secret' and 'secretFile' should not both be set";
+      };
+
+    };
+  };
+
+  ###### implementation
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ pkgs.ipfs-cluster ];
+    environment.variables.IPFS_CLUSTER_PATH = cfg.dataDir;
+
+    systemd.tmpfiles.rules =
+      [ "d '${cfg.dataDir}' - ${cfg.user} ${cfg.group} - -" ];
+
+    systemd.packages = [ pkgs.ipfs-cluster pkgs.coreutils ];
+
+    systemd.services.ipfs-cluster-init = {
+      path = [ "/run/wrappers" pkgs.ipfs-cluster ];
+      environment = {
+        IPFS_CLUSTER_PATH = cfg.dataDir;
+        CLUSTER_SECRET = if (cfg.secret != null) then
+          cfg.secret elseif (secretFile != null) readFile cfg.secretFile
```

Comment:
This invocation of `readFile` will pull your secrets file into the nix store, which we don't want.